### ISSUE TITLE
Support re-weighting of subjects

### ISSUE THREAD

**Issue by kamilest** [kamilest/meds-evaluation#7]
For resampled metrics, the implementation currently samples (with replacement) the same number of events per subject in order to represent each subject equally. This is inefficient and possibly leads to higher variance estimates.

This should be eventually replaced with re-weighting of subjects where all events of a subject is used but the contributions of each patient should be weighted appropriately so that each subject is represented equally. 

This might be more challenging for some of the metrics (e.g. AUROC), where we should clearly specify how to do this correctly.

cc: @mmcdermott @EthanSteinberg

**Comment by EthanSteinberg** [comment]
Sounds good.

One comment is that this is trivial for AUROC, simply provide the sample_weight option to the calculator: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score

**Comment by mmcdermott** [comment]
I think @kamilest's comment about how this may be more challenging for AUROC is reflecting the fact that the following three things are (I think) different:
  1. Compute AUROC over all the samples without regard to how many samples a single patient has. (this you can do trivially with `roc_auc_score` by not specifying sample weights).
  2. Compute AUROC such that the FPR and TPR calculations for each patient are weighted by a set of sample weights such that each patient gets, in total, the same aggregate weight (this you can do trivially with `roc_auc_score` by specifying sample weights).
  3. Compute AUROC for each patient individually, with only the predictions for that patient, then average all of those individual AUROCs together. (this is not trival to compute afaik).
  
These are all different, and they have different sensible use cases.
1. This is used when your deployment / decision scenario is about contrasting single examples against other examples.
2. This I don't know when or if it should be used.
3. This should be used when your deployment / decision scenario is about contrasting examples for a given patient/group/subject against other examples only within that group. The most compelling example I have for this is something like predicting COVID rates for the next 30 days for geographic regions across a big dataset of different regions. Each region will usually have a fully independent decision process, budget allocation process, etc., than other regions, so it may not matter if I could appropriately distinguish when a given month for county A either appropriately or inappropriately has a score higher than a given month for county B. But, within county A, being able to use the score to dynamically adjust one's response over time might be much more useful, so a per-county AUROC may be more relevant than an overall AUROC.

**Comment by EthanSteinberg** [comment]
@mmcdermott It's interesting that you write that you don't know when or if 2 should be used, but 2 is what you currently have implemented with the sampling code.

**Comment by mmcdermott** [comment]
Oh, no, I know, Ethan. This is a new revelation for me, and I am now no
longer trying to advocate for the resampling or reweighing approach for
AUC. For accuracy and other simpler point wise metrics I think 2 and 3 are
identical, so the approach still makes sense there, but for AUC and other
non-pointwise metrics they differ.

**Comment by EthanSteinberg** [comment]
I think 2 actually makes a lot of sense here. I think it's equivalent to saying making a mistake on a given patient has equal cost, no matter how many labels that patient has. So patients with 100 labels have the same cost of mistakes as patients with 1 label.

You can think about it as very similar to a lot of fairness criteria to a certain extent in that you are modifying your utility function to make sure that no patients are ignored.

**Comment by mmcdermott** [comment]
I think this can probably be de-prioritized in the immediate future and we could even just focus on providing per-sample evaluations by default or sample-reweighted evaluations by default (and I'd recommend the former). I'm not suggesting a better approach wouldn't be preferred here, just that it is not high priority at this time.
